contracts/mystery-box/src/lib.rs

```diff
@@ -0,0 +1,142 @@
+    + rewards::RewardsModule
+    + permissions_module::PermissionsModule
+    + events::EventsModule
+{
+    #[init]
+    fn init(&self, mystery_box_token_id: TokenIdentifier) {
+        require!(
+            mystery_box_token_id.is_valid_esdt_identifier(),
+            "Invalid token ID"
+        );
+        self.mystery_box_token_identifier()
+            .set(mystery_box_token_id);
+
+        let caller = self.blockchain().get_caller();
+        self.add_admin(&caller);
+    }
+
+    #[endpoint(createMysteryBox)]
+    fn create_mystery_box(
+        &self,
+        receiver: ManagedAddress,
+        mystery_box_setup: ManagedVec<Reward<Self::Api>>,
+    ) -> EsdtTokenPayment<Self::Api> {
+        let caller = self.blockchain().get_caller();
+        self.require_admin(&caller);
+
+        let mystery_box_token_id = self.mystery_box_token_identifier().get();
+
+        let mut accumulated_percentage = 0u64;
+
+        for mystery_box_reward in mystery_box_setup.into_iter() {
+            accumulated_percentage += mystery_box_reward.percentage_chance;
+            self.check_reward_validity(&mystery_box_reward);
+        }
+        require!(
+            accumulated_percentage == MAX_PERCENTAGE,
+            "The total percentage must be 100%"
+        );
+
+        let mystery_box_token_nonce = self.send().esdt_nft_create_compact(
+            &mystery_box_token_id,
+            &BigUint::from(NFT_AMOUNT),
+            &mystery_box_setup,
+        );
+
+        let user_payment = EsdtTokenPayment::new(
+            mystery_box_token_id,
+            mystery_box_token_nonce,
+            BigUint::from(1u64),
+        );
+        self.send().direct_esdt(
+            &receiver,
+            &user_payment.token_identifier,
+            user_payment.token_nonce,
+            &user_payment.amount,
+        );
+
+        self.emit_create_mystery_box_event(&receiver, &mystery_box_setup);
+
+        user_payment
+    }
+
+    #[payable("*")]
+    #[endpoint(openMysteryBox)]
+    fn open_mystery_box(&self) {
+        let mystery_box_token_id = self.mystery_box_token_identifier().get();
+        let payment = self.call_value().single_esdt();
+        require!(
+            payment.token_identifier == mystery_box_token_id,
+            "Bad payment token"
+        );
+        let mystery_box_attributes: ManagedVec<Reward<Self::Api>> = self
+            .blockchain()
+            .get_esdt_token_data(
+                &self.blockchain().get_sc_address(),
+                &payment.token_identifier,
+                payment.token_nonce,
+            )
+            .decode_attributes();
+
+        let current_epoch = self.blockchain().get_block_epoch();
+        let cooldown_epoch = self.cooldown_epoch().get();
+        let mut active_cooldown = true;
+        let mut winning_reward = Reward::default();
+        while active_cooldown {
+            winning_reward = self.get_winning_reward(&mystery_box_attributes);
+            active_cooldown =
```

Comment:
What do you want with this while code? 
The SFT of misteryBox should contain an creation epoch. And you should check only that currentepoch - creationEpoch > cooldownPeriod